### PR TITLE
[RDSS-1836] Fix always passing tests, and version field specified to be SemVer

### PR DESCRIPTION
**Pull request by michalc**

### CODE REVIEW

File: schemas/header.json [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+              "machineId",
+              "machineAddress",
+              "timestamp"
+            ],
+            "additionalProperties": false
+          }
+        },
+        "version": {
+          "type": "string",
+          "pattern": "^(([0-9]|[1-8][0-9]|9[0-9]|[1-8][0-9]{2}|9[0-8][0-9]|99[0-9]|[1-8][0-9]{3}|9[0-8][0-9]{2}|99[0-8][0-9]|999[0-9])(\\.|$)){3}"
```
Comment (danielgrant) [inline schemas/header.json]: I have tested various SemVer-like version strings against the pattern here and there are some issues:

- It accepts `1.2.3.`, which should fail
- It accepts `1.2.3.4`, which should fail
- It doesn't accept sections 9 and 10 of SemVer 2.0.0, which should pass.

Comment (michalc) [reply schemas/header.json]: Ah will check 1.2.3.4.

However, sections 9 and 10 should _not_ be accepted according to the ticket. I commented on this in the commit message.

**Comment by michalc** [comment]
@danielgrant 

(Repeating other comments, so they're all in the one place)

1. Fixed (and much simplified) the regex, so it should now forbid 1.2.3.4 (I originally used a generator when actually, doing it myself was much simpler :-) )
2. The regex does not allow 1.2.3-something or 1.2.3+else, sections 9 and 10 from semver.org. I don't think the ticket requests this: only that the versions be in the form 1.2.3 (and I think this was discussed during a call). This is explained in the commit message.
3. I've re-ordered the commits so the fixing of the test is first in the history: essentially so the changes regarding the regex are more easily found from the future.

However, I'm not sure what tests were not working properly: it's not clear to me by the commit message. Let me know, and I will amend the message? (or feel free to as well)

Also, I wonder if it's good to have some messages that should _not_ validate as tests?

**Comment by michalc** [comment]
I have a few comments on the process, if that's ok?

> This shouldn't change until it's master, otherwise it implies that what is in the develop branch is a release.

> Once develop has been merged into master, a branch is taken to update the version to a release version. Once merged, a tag and release is taken. master is then merged into develop, and the process repeated for the develop branch, but this time applying a -SNAPSHOT version.

I've been pondering these, but wonder if this process can be simplified? It does also allow for a bit of a race condition when code is in master that is backwards-incompatible but the version has not been bumped? Having something in master that doesn't have an accurate version, if using SemVer, seems quite bad?

I wonder whether the version bumps should just be done in the same branches as the code changes that led to them. Then releases are just tags/releases as now.

However, I am coming from quite a continual deployment point of view. Essentially planning so every change _is_ releasable. If planning for changes to _not_ be releasable, then perhaps the current process is appropriate. However, so far what I've worked on in the past couple of weeks, changes have been releasable.

However, perhaps this is all a discussion for another time. I'm mostly just wanting to sow-the-seeds of something simpler in future for people to mull over... (or for me to mull over counter-arguments :-) )

**Comment by michalc** [comment]
@danielgrant 

> I take your point that this process could be improved through a CI / CD set up, but that's just something we don't have in place at this point in time.

I have more points in response, but perhaps another time!

Just to check, would be good to know which tests are the ones being fixed by the changes you added, and how they were fixed. Let me know? I'm not that keen adding something that I don't understand now, and that the git message doesn't make clear, to the git history.

**Comment by michalc** [comment]
(And I will update the regex to be SemVer compatible)

Also: should I add some tests that checks that _invalid_ messages fail?

**Comment by danielgrant** [comment]
> Just to check, would be good to know which tests are the ones being fixed by the changes you added, and how they were fixed.

The way the `message_header.json` schema file was being used was resulting in the `header.json` file not being validated. This has been refactored out into its own dedicated JSON schema file (in the `schemas/` folder) and the `header` object defined within is now referenced using the qualified URI of that `header` object. This now causes the required validation to occur.

> Also: should I add some tests that checks that invalid messages fail?

I can see the value in that but I would suggest we add a ticket to the backlog to pick that up. I don't want to start introducing extra scope in an already defined ticket.

**Comment by michalc** [comment]
@danielgrant 

> The way the message_header.json schema file was being used was resulting in the header.json file not being validated. This has been refactored out into its own dedicated JSON schema file (in the schemas/ folder) and the header object defined within is now referenced using the qualified URI of that header object. This now causes the required validation to occur.

Thanks!

> I can see the value in that but I would suggest we add a ticket to the backlog to pick that up. I don't want to start introducing extra scope in an already defined ticket.

That's fine.

(Er, although I might have a suggestion for a change in process again: I am used to writing tests for a ticket as part of that ticket: it is implicitly part of its scope. Again, this is from semi-TDD point of view, which admittedly may not be appropriate in all cases)

**Comment by michalc** [comment]
@danielgrant 

I've updated the regex to what I _think_ is SemVer 2.0.0 . It's getting to be quite a long regex! Note your comment

  > -SNAPSHOT for example is not compatible

This _is_allowed by SemVer I think, so I might misunderstand your comment?

I've also amended the comment for your commit. Let me know if it doesn't make sense.

**Comment by sevein** [comment]
@michalc the regex looks pretty good. I've only noticed that the major, minor and patch components are being captured twice. Is that necessary?

image

I've removed the groups as follows:

    "^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(-([0-9A-Za-z-]+\\.)*[0-9A-Za-z-]+)?(\\+([0-9A-Za-z-]+\\.)*[0-9A-Za-z-]+)?$"

image

**Comment by michalc** [comment]
@sevein Ah yes, I think that was unnecessary! I have replaced it with your simplified form.

**Pull request merged**